Prevent `RequestEncoder#encode_params` to parse falsey params

al3rez opened the pull request:
Fixes #33388 - When a `get` method called with `as: :json` and `params: nil`
(explicitly or implicitly) `RequestEncoder#encode_params` converts it
into a `null` value which includes a unexpected `null=` query string
into request URL. From now on `RequestEncoder#encode_params` checks
whether `params` is nil or not otherwise returns.

azbshiri commented:
@rafaelfranca

bogdanvlviv commented on actionpack/test/controller/integration_test.rb:
```diff
@@ -4,6 +4,7 @@
 require "controller/fake_controllers"
 require "rails/engine"
 
+
```
Could you remove this extra line?

azbshiri replied:
done

bogdanvlviv commented on actionpack/lib/action_dispatch/testing/request_encoder.rb:
```diff
@@ -34,7 +34,7 @@ def accept_header
 
     def encode_params(params)
-      @param_encoder.call(params)
+      @param_encoder.call(params) unless params.nil?
```
I think it shouldn't be fixed in `actionpack/lib/action_dispatch/testing/integration.rb` See diff:
```diff
diff --git a/actionpack/lib/action_dispatch/testing/integration.rb b/actionpack/lib/action_dispatch/testing/integration.rb
index 7637febd1c..9c6075ad0c 100644
--- a/actionpack/lib/action_dispatch/testing/integration.rb
+++ b/actionpack/lib/action_dispatch/testing/integration.rb
@@ -233,7 +233,7 @@ def process(method, path, params: nil, headers: nil, env: nil, xhr: false, as: n

         request_env = {
           :method => method,
-          :params => request_encoder.encode_params(params),
+          :params => params.nil? ? nil : request_encoder.encode_params(params),
```

azbshiri replied:
done

kaspth replied:
@bogdanvlviv why shouldn't it be fixed in here? Isn't it a special case of the request encoder to handle?

I'd write it as `@param_encoder.call(params) if params` though.

Should we update some documentation too? E.g. to say that a request encoder will only be asked to encode params when they aren't nil or similar?

bogdanvlviv replied:
Yeah, probably would be better just to write `@param_encoder.call(params) if params` in  `ActionDispatch::RequestEncoder`, Since it is private api we haven't written any documentation.
By the way if

```ruby
get "/foos_json", as: :json, params: false
request.url # => http://www.example.com/foos_json?false
```
condition `if params` would fix it too.

@azbshiri My apologies for confusing. Could you apply `@param_encoder.call(params) if params`, and update the commit message accordance this change?

azbshiri replied:
done

azbshiri replied:
I didn't commit and push the changes of Yesterday, it's ready to be merged.

bogdanvlviv commented:
@azbshiri Could you squash commits  into one commit?

azbshiri commented:
done

silva96 commented on actionpack/lib/action_dispatch/testing/integration.rb:
```diff
@@ -233,7 +233,7 @@ def process(method, path, params: nil, headers: nil, env: nil, xhr: false, as: n
         request_env = {
           :method => method,
-          :params => request_encoder.encode_params(params),
+          :params => params.nil? ? nil : request_encoder.encode_params(params),
```
I would remove it from here since this is a specific usage of the encode_params which already handles it.

azbshiri replied:
as i mentioned in tsolar reply that was the case, but i forgot to commit.

tsolar commented on actionpack/lib/action_dispatch/testing/integration.rb:
```diff
@@ -233,7 +233,7 @@ def process(method, path, params: nil, headers: nil, env: nil, xhr: false, as: n
         request_env = {
           :method => method,
-          :params => request_encoder.encode_params(params),
+          :params => params.nil? ? nil : request_encoder.encode_params(params),
```
if `encode_params` returns `nil` if `params` not present, then this conditional is not necessary

azbshiri replied:
@tsolar yeah it was the case, but i didn't commit yesterday.

Commit:
Prevent `RequestEncoder#encode_params` to parse falsey params

When a `get` method called with `as: :json` and `params: nil` or
`params: false` (explicitly or implicitly)
`RequestEncoder#encode_params` converts it into a `null` or `false`
value which includes a unexpected `null=` or `false` query string into
request URL. From now on `RequestEncoder#encode_params` checks whether
`params` is nil or not otherwise returns.

Move down `nil` conversion guard


Update CHANGELOG.md

kaspth commented:
Thanks!
